src/ILCompiler.Compiler/src/Compiler/TypePreinit.cs

```diff
@@ -1698,21 +1836,86 @@ public override void WriteFieldData(ref ObjectDataBuilder builder, FieldDesc fie
             }
         }
 
-        private abstract class ReferenceTypeValue : Value, ISerializableReference
+        private abstract class ReferenceTypeValue : Value
         {
-            protected readonly TypeDesc _type;
+            public TypeDesc Type { get; }
 
-            protected ReferenceTypeValue(TypeDesc type) { _type = type; }
+            protected ReferenceTypeValue(TypeDesc type) { Type = type; }
 
             public override bool Equals(Value value)
             {
                 return this == value;
             }
+        }
+
+        private class DelegateInstance : ReferenceTypeValue, ISerializableReference
+        {
+            private readonly MethodDesc _methodPointed;
+            private readonly ForeignTypeInstance _firstParameter;
+
+            public DelegateInstance(TypeDesc delegateType, MethodDesc methodPointed, ForeignTypeInstance firstParameter)
+                : base(delegateType)
+            {
+                _methodPointed = methodPointed;
+                _firstParameter = firstParameter;
+            }
+
+            public virtual void WriteContent(ref ObjectDataBuilder builder, ISymbolNode thisNode, NodeFactory factory)
+            {
+                Debug.Assert(_methodPointed.Signature.IsStatic == (_firstParameter == null));
+                Debug.Assert(!_methodPointed.IsUnmanagedCallersOnly);
```

Comment:
I assume that this is counterpart to the `InitializeReversePInvokeThunk` hack `DelegateCreationInfo Create`.

Do we really need that hack? I believe that everything is still going to work fine even if we delete the `IsUnmanagedCallersOnly` check in  `DelegateCreationInfo Create`.

Reply:
Isn't this needed to make `AddrOf` intrinsics kind of work with delegates that point to `UnmanagedCallersOnly` methods? Those delegates are supposed to be useless otherwise. I think we can delete that once we pick up a C# compiler that has function pointers and align with whatever CoreCLR is doing for delegates pointing to UnmanagedCallersOnly methods (throw at delegate creation time?).

Reply:
Yes, these delegates are useless. I do not see why we need to special case them. It can be regular useless delegate (it is what happens in CoreCLR when you create delegate to `UnmanagedCallersOnly`).

Reply:
Looks like CoreCLR blocks their creation. We can't block their creation just yet because we need their instances for `AddrOf`. If we allow them to be created like regular delegates, I assume bad things will happen if someone calls them, no?

I really should move this assert and make this a reason to bail the preinitialization.

Reply:
CoreCLR used to block their creation, but the block was not consistent. We have removed the block in https://github.com/dotnet/runtime/pull/35763 . We just fail fast during the invocation now.

Reply:
I see, I tried it with the public preview that probably doesn't have the fix and still blocks.